vm/actor/src/builtin/miner/deadline_state.rs

```diff
@@ -0,0 +1,998 @@
+
+            // Save partition back.
+            partitions.set(partition_idx, partition)?;
+
+            // Record deadline -> partition mapping so we can later update the deadlines.
+            for sector in partition_new_sectors {
+                if let Some(partition_update) =
+                    partition_deadline_updates.get_mut(&sector.expiration)
+                {
+                    if partition_update.last() != Some(&partition_idx) {
+                        partition_update.push(partition_idx);
+                    }
+                }
+            }
+        }
+
+        // Save partitions back.
+        self.partitions = partitions.flush()?;
+
+        // Next, update the expiration queue.
+        let mut deadline_expirations =
+            BitFieldQueue::new(store, &self.expirations_epochs, quant)
+                .map_err(|e| format!("failed to load expiration epochs: {:?}", e))?;
+        deadline_expirations
+            .add_many_to_queue_values(&partition_deadline_updates)
+            .map_err(|e| {
+                ActorError::downcast_wrap(e, "failed to add expirations for new deadlines")
+            })?;
+        self.expirations_epochs = deadline_expirations.amt.flush()?;
+
+        Ok(new_power)
+    }
+
+    pub fn pop_early_terminations<BS: BlockStore>(
+        &mut self,
+        store: &BS,
+        max_partitions: u64,
+        max_sectors: u64,
+    ) -> Result<(TerminationResult, /* has more */ bool), Box<dyn StdError>> {
+        let mut partitions = self.partitions_amt(store)?;
+
+        let mut partitions_finished = Vec::<u64>::new();
+        let mut result = TerminationResult::new();
+
+        for i in self.early_terminations.iter() {
+            let partition_idx = i as u64;
+
+            let mut partition = match partitions
+                .get(partition_idx)
+                .map_err(|e| format!("failed to load partition {}: {:?}", partition_idx, e))?
+            {
+                Some(partition) => partition,
+                None => {
+                    partitions_finished.push(partition_idx);
+                    continue;
+                }
+            };
+
+            // Pop early terminations.
+            let (partition_result, more) = partition
+                .pop_early_terminations(store, max_sectors - result.sectors_processed)
+                .map_err(|e| {
+                    ActorError::downcast_wrap(e, "failed to pop terminations from partition")
+                })?;
+
+            result += partition_result;
+
+            // If we've processed all of them for this partition, unmark it in the deadline.
+            if !more {
+                partitions_finished.push(partition_idx);
+            }
+
+            // Save partition
+            partitions
+                .set(partition_idx, partition)
+                .map_err(|e| format!("failed to store partition {}: {:?}", partition_idx, e))?;
+
+            if !result.below_limit(max_partitions, max_sectors) {
```

Comment:
```suggestion
            if result.below_limit(max_partitions, max_sectors) {
```

Reply:
I think my logic is correct here, the Go code returns `nil` (i.e. continues iteration) when `below_limit` returns false